pkg/endpoint/bpf.go

```diff
@@ -459,34 +459,53 @@ func (e *Endpoint) removeOldRedirects(owner Owner, desiredRedirects map[string]b
 		}
 }
 
-// regenerateBPF rewrites all headers and updates all BPF maps to reflect the
-// specified endpoint.
-// Must be called with endpoint.Mutex not held and endpoint.BuildMutex held.
-// Returns the policy revision number when the regeneration has called, a
-// boolean if the BPF compilation was executed and an error in case of an error.
-func (e *Endpoint) regenerateBPF(owner Owner, currentDir, nextDir string, regenContext *RegenerationContext) (revnum uint64, compiled bool, reterr error) {
-	var (
-		err                 error
-		compilationExecuted bool
-	)
+// datapathRegenerationContext contains information related to regenerating the
+// datapath (BPF, proxy, etc.).
+type datapathRegenerationContext struct {
+	bpfHeaderfilesHash    string
+	epInfoCache           *epInfoCache
+	bpfHeaderfilesChanged bool
+	proxyWaitGroup        *completion.WaitGroup
+	ctCleaned             chan struct{}
```

Comment:
The attractive thing about storing this field within the `datapathRegenerationContext` is that we can refactor it as needed instead of having a large amount of state included in a single function. In the future I'd like to break up `runPreCompilationSteps` to be smaller and more modular if possible.

Regarding your last question, I'm not sure. I'd have to dig through the `git log`.